lib/view_model/access_control.rb

```diff
@@ -0,0 +1,126 @@
+
+  # Check that the record is eligible to be changed in its current state, in the
+  # given context. During deserialization, this must be called before any edits
+  # have taken place (thus checking against the initial state of the viewmodel),
+  # and if edit is denied, an error must be raised if an edit is later
+  # attempted. To be overridden by viewmodel implementations.
+  def editable_check(view, deserialize_context:)
+    Result::DENY
+  end
+
+  # Check that the attempted changes to this record are permitted in the given
+  # context. During deserialization, this must be called once all edits have
+  # been attempted. To be overridden by viewmodel implementations.
+  def valid_edit_check(view, deserialize_context:, changes:)
+    Result::DENY
+  end
+
+  # Implementations of deserialization that will potentially make changes to the
+  # viewmodel must call this on the viewmodel to save the initial `editable_check`
+  # value and optional exception before attempting to apply their changes.
+  def initial_editability(view, deserialize_context:)
+    return nil if ineligible(view)
+    editable_check(view, deserialize_context: deserialize_context)
+  end
+
+  # Implementations of serialization and deserialization must call this
+  # whenever a viewmodel is visited during serialization or deserialization.
+  def visible!(view, context:)
+    return if ineligible(view)
+
+    result = visible_check(view, context: context)
+
+    access_check!(result) do
+      message =
+        if context.is_a?(ViewModel::DeserializeContext)
+          "Attempt to deserialize into forbidden viewmodel '#{view.class.view_name}'"
+        else
+          "Attempt to serialize forbidden viewmodel '#{view.class.view_name}'"
+        end
+
+      ViewModel::AccessControlError.new(message, view.blame_reference)
+    end
+  end
+
+  # Implementations of deserialization must call this when they know what
+  # changes are to be made to the viewmodel. For viewmodels with transactional
+  # backing models, the changes may be made in advance to give the edit checks
+  # the opportunity to compare values. Must be called with the saved
+  # `initial_editability` value if changes have been made.
+  def editable!(view, initial_editability: nil, deserialize_context:, changes:)
+    return if ineligible(view)
+
+    initial_editability ||= editable_check(view, deserialize_context: deserialize_context)
+
+    result = initial_editability.merge do
+      valid_edit_check(view, deserialize_context: deserialize_context, changes: changes)
+    end
+
+    access_check!(result) do
+      ViewModel::AccessControlError.new(
+        "Illegal edit to viewmodel '#{view.class.view_name}'",
+        view.blame_reference)
+    end
+  end
+
+  private
+
+  def ineligible(view)
+    # ARVM synthetic views are considered part of their association and as such
+    # are not edit checked.
+    # TODO: Do we want to make this a more generic scheme so that any VM type
```

Comment:
I don't currently expect this to be exposed to the application. Plan for enumerated types visibility is define an AccessControl::Tree that specifies them as visible/immutable and include it as relevant as an `ancestor`.